### PR TITLE
Task/14279 Implement SRS submission process

### PR DESCRIPTION
**Pull request by 30mar**
## Description
**Included**
Add the logic for device token and APIToken.
Add request for otp auth
Add SRSService.
**Missing**: 
Build errors from the network request need to be adapted to the new restService
final submittion call

## Link to Jira
https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-14279

### CODE REVIEW

File: src/xcode/ENA.xcworkspace/xcshareddata/swiftpm/Package.resolved [hunk 1]
```diff
@@ -82,6 +82,15 @@
           "version": null
         }
       },
+      {
+        "package": "FMDB",
```
Comment (naveeddotio) [inline src/xcode/ENA.xcworkspace/xcshareddata/swiftpm/Package.resolved]: Please revert package changes, we can't update without rigorously testing.

File: src/xcode/ENA/ENA/Resources/Localization/de.lproj/Localizable.strings [hunk 2]
```diff
@@ -727,6 +727,10 @@ Bei ausgeschalteter Hintergrundaktualisierung müssen Sie die App täglich aufru
 
 "SRS_ConsentScreen_header_section1" = "Andere warnen";
 
+"SRSConsentScreen_HelpOthers" = "Helfen Sie mit, indem Sie andere warnen, denen Sie begegnet sind!.";
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Resources/Localization/de.lproj/Localizable.strings]: Please revert these changes as well

File: src/xcode/ENA/ENA/Resources/Localization/de.lproj/Localizable.strings [hunk 3]
```diff
@@ -727,6 +727,10 @@ Bei ausgeschalteter Hintergrundaktualisierung müssen Sie die App täglich aufru
...
+"SRSConsentScreen_title_description2" = "Wenn Ihr Corona-Test positiv ist, können Sie Nutzer über die App warnen. Es wird nur innerhalb Deutschlands gewarnt.
```
Comment (KathrinLuecke) [inline src/xcode/ENA/ENA/Resources/Localization/de.lproj/Localizable.strings]: Where does this text come from? In Figma I see another text for the first bullet on the consent screen:
"Wenn Sie positiv getestet wurden, können Sie Ihre Mitmenschen über die App warnen.
Die Warnung nach einem positiven Selbsttest oder mit einem Testergebnis, das nicht in der App vorliegt, erfordert die Prüfung der Echtheit der App."
https://www.figma.com/file/KuDEcMSREg3VhZRhC3FXTY/COVID19_Master_2.25?node-id=53209%3A42588&t=B9G9THS0dDhSiyTy-4

Comment (naveeddotio) [reply src/xcode/ENA/ENA/Resources/Localization/de.lproj/Localizable.strings]: The new changes are adapted with this PR

https://github.com/corona-warn-app/cwa-app-ios/pull/4896

File: src/xcode/ENA/ENA/Resources/Localization/de.lproj/Localizable.strings [hunk 4]
```diff
@@ -499,7 +499,7 @@ Bei ausgeschalteter Hintergrundaktualisierung müssen Sie die App täglich aufru
 
 "OnboardingInfo_alwaysStayInformedPage_title" = "Warnungen erhalten, Risiken erkennen";
 
-"OnboardingInfo_alwaysStayInformedPage_boldText" = "Die App kann Ihnen automatisch Mitteilungen senden und Sie unter anderem über Ihren Risikostatus oder die Gültigkeit Ihrer Zertifikate informieren. Passen Sie dafür die Mitteilungseinstellungen an.";
+"OnboardingInfo_alwaysStayInformedPage_boldText" = "Die App kann Ihnen automatisch Mitteilungen senden und Sie unter anderem über Ihren Risikostatus oder die Gültigkeit Ihrer Zertifikate informieren. Passen Sie dafür die Mitteilungseinstellungen an.";
```
Comment (puneetmahali1) [inline src/xcode/ENA/ENA/Resources/Localization/de.lproj/Localizable.strings]: Is there any change in the text?

File: src/xcode/ENA/ENA/Source/HTTPClient/Resources/KeySubmission/SRSKeySubmissionResource.swift [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+	}
+}
+
+struct SRSKeySubmissionResource: Resource {
+
+	init(
+		payload: SubmissionPayload,
+		srsOtp: String,
+		isFake: Bool = false,
+		// submissionType: SRSSubmissionType,
```
Comment (puneetmahali1) [inline src/xcode/ENA/ENA/Source/HTTPClient/Resources/KeySubmission/SRSKeySubmissionResource.swift]: Do we need this commented line?

File: src/xcode/ENA/ENA/Resources/Localization/de.lproj/Localizable.strings (hunk 4 above)
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Resources/Localization/de.lproj/Localizable.strings:502]: It could be a wrong character, happened to me once before.

File: src/xcode/ENA/ENA/Source/HTTPClient/LocatorsNotYetUsed/KeySubmission.swift [hunk 6]
```diff
@@ -11,22 +11,43 @@ extension Locator {
...
-				"cwa-authorization": payload.tan,
-				"cwa-fake": fake,
-				"cwa-header-padding": fakePadding
-			]
-		)
+		
+		// we have an extra special key in the header to differentiate SRS from normal submission
+		// for that we have the optional srsOtp argument in the function and based on it we
+		// either add the srs header or the normal submission header
+		if let srsOtp = srsOtp {
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/HTTPClient/LocatorsNotYetUsed/KeySubmission.swift:24]: 👍

File: src/xcode/ENA/ENA/Source/HTTPClient/LocatorsNotYetUsed/authorizeOtpSrs.swift [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+//
+
+import Foundation
+
+extension Locator {
+    
+    // send:	ProtoBuf SAP_Internal_Ppdd_SRSOneTimePasswordRequestIOS
+    // receive:	JSON
+    // type:	default
+    // comment:
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/HTTPClient/LocatorsNotYetUsed/authorizeOtpSrs.swift]: Missing comment?

File: src/xcode/ENA/ENA/Source/HTTPClient/Resources/KeySubmission/SRSKeySubmissionResource.swift (hunk 5 above)
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/HTTPClient/Resources/KeySubmission/SRSKeySubmissionResource.swift]: could be removed?

File: src/xcode/ENA/ENA/Source/HTTPClient/Resources/KeySubmission/SRSKeySubmissionResource.swift [hunk 8]
```diff
@@ -0,0 +1,85 @@
+//
+// 🦠 Corona-Warn-App
+//
+
+import Foundation
+
+enum SRSKeySubmissionResourceError: LocalizedError, Equatable {
+	case invalidPayloadOrHeaders
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/HTTPClient/Resources/KeySubmission/SRSKeySubmissionResource.swift]: Could be treated as singular

```suggestion
	case invalidPayloadOrHeader
```

File: src/xcode/ENA/ENA/Source/HTTPClient/Resources/KeySubmission/SRSKeySubmissionResource.swift [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+	case serverError(Int)
+	
+	var errorDescription: String? {
+		switch self {
+		case let .serverError(code):
+			return "\(AppStrings.ExposureSubmissionError.other)\(code) \(AppStrings.ExposureSubmissionError.otherend)"
+		case .invalidPayloadOrHeaders:
+			return "\(AppStrings.ExposureSubmissionError.errorPrefix) - Received an invalid payload or headers."
+		case .invalidOtp:
+			// to.do update text for invalid otp
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/HTTPClient/Resources/KeySubmission/SRSKeySubmissionResource.swift]: Still not available?

File: src/xcode/ENA/ENA/Source/HTTPClient/Resources/KeySubmission/SRSKeySubmissionResource.swift [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+		self.type = .default
+		self.receiveResource = EmptyReceiveResource()
+		self.trustEvaluation = trustEvaluation
+		
+		self.sendResource = ProtobufSendResource(
+			SAP_Internal_SubmissionPayload.with {
+				$0.requestPadding = payload.exposureKeys.submissionPadding
+				$0.keys = payload.exposureKeys
+				$0.checkIns = payload.checkins
+				// Consent needs always set to be false with SRS
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/HTTPClient/Resources/KeySubmission/SRSKeySubmissionResource.swift:47]: 👍

File: src/xcode/ENA/ENA/Source/HTTPClient/Resources/OTPAuthorizationForSRS/OTPAuthorizationForSRSResource.swift [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+		)
+	) {
+		let ppacIos = SAP_Internal_Ppdd_PPACIOS.with {
+			$0.apiToken = ppacToken.apiToken
+			$0.deviceToken = ppacToken.deviceToken
+		}
+		let payload = SAP_Internal_Ppdd_SRSOneTimePassword.with {
+			$0.otp = otpSRS
+		}
+        self.sendResource = ProtobufSendResource(
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/HTTPClient/Resources/OTPAuthorizationForSRS/OTPAuthorizationForSRSResource.swift]: wrong indentation or GH issue?

File: src/xcode/ENA/ENA/Source/HTTPClient/Resources/OTPAuthorizationForSRS/OTPAuthorizationForSRSResource.swift [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+            return .noNetworkConnection
+        case .unexpectedServerError(let statusCode):
+            switch statusCode {
+            case 400, 401, 403:
+                return otpAuthorizationFailureHandler(for: responseBody, statusCode: statusCode)
+            case 500:
+                Log.error("Failed to get authorized OTP - 500 status code", log: .api)
+                return .otherServerError
+            default:
+                Log.error("Failed to authorize OTP - response error", log: .api)
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/HTTPClient/Resources/OTPAuthorizationForSRS/OTPAuthorizationForSRSResource.swift]: could we print statusCode after response error: and not in next line?

File: src/xcode/ENA/ENA/Source/HTTPClient/Resources/OTPAuthorizationForSRS/OTPAuthorizationForSRSResource.swift [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+        default:
+            return .otherServerError
+        }
+    }
+	
+	// MARK: - Private
+	
+	private func otpAuthorizationFailureHandler(for response: Data?, statusCode: Int) -> OTPAuthorizationError? {
+		guard let responseBody = response else {
+			Log.error("Failed to get authorized OTP - no 200 status code", log: .api)
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/HTTPClient/Resources/OTPAuthorizationForSRS/OTPAuthorizationForSRSResource.swift]: Same here

File: src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/ExposureSubmissionCoordinator.swift [hunk 14]
```diff
@@ -473,52 +475,52 @@ class ExposureSubmissionCoordinator: NSObject, RequiresAppDependencies {
...
-
-		let topBottomContainerViewController = TopBottomContainerViewController(
-			topController: srsConsentViewController,
-			bottomController: FooterViewController(footerViewModel)
-		)
-
-		return topBottomContainerViewController
-	}
-	
+    private func makeSRSConsentScreen(srsFlowType: SRSFlowType) -> UIViewController {
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/ExposureSubmissionCoordinator.swift:478]: GH issue?

File: src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/ExposureSubmissionCoordinator.swift [hunk 15]
```diff
@@ -1685,6 +1687,8 @@ class ExposureSubmissionCoordinator: NSObject, RequiresAppDependencies {
 		self.model.submitExposure(
 			isLoading: isLoading,
 			onSuccess: { [weak self] in
+				self?.store.mostRecentKeySubmissionDate = Date()
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/ExposureSubmissionCoordinator.swift:1690]: 👍

File: src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/ExposureSubmissionCoordinatorModel.swift [hunk 16]
```diff
@@ -184,22 +186,42 @@ class ExposureSubmissionCoordinatorModel {
...
+						Log.info("Consent Not Given", log: .ui)
+						onSuccess()
+						
+					case .some(let error):
+						Log.error("error: \(error.localizedDescription)", log: .api)
+						onError(error)
+					}
+				}
+			case .failure(let srsError):
+				// to.do present srsError "APITOKEN or OTP auth failed"
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/ExposureSubmissionCoordinatorModel.swift]: Should we create a separate task and collect all the to.dos left after this PR?

File: src/xcode/ENA/ENA/Source/Services/Exposure Submission/ExposureSubmissionService.swift [hunk 17]
```diff
@@ -150,7 +150,80 @@ class ENAExposureSubmissionService: ExposureSubmissionService {
...
 		}
 	}
+	/// This method submits the SRS exposure keys. Additionally, after successful completion,
+	/// the timestamp of the key submission is updated.
+	func submitSRSExposure(
+		submissionType: SRSSubmissionType,
+		srsOTP: String,
+		completion: @escaping (_ error: ExposureSubmissionServiceError?) -> Void
+	) {
+		Log.info("Started exposure submission...", log: .api)
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/Services/Exposure Submission/ExposureSubmissionService.swift]:
```suggestion
		Log.info("Started SRS exposure submission...", log: .api)
```

File: src/xcode/ENA/ENA/Source/Services/Exposure Submission/ExposureSubmissionService.swift [hunk 18]
```diff
@@ -150,7 +150,80 @@ class ENAExposureSubmissionService: ExposureSubmissionService {
...
+
+		guard let keys = temporaryExposureKeys else {
+			Log.info("Cancelled submission: No temporary exposure keys to submit.", log: .api)
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/Services/Exposure Submission/ExposureSubmissionService.swift]:
```suggestion
			Log.info("Cancelled SRS exposure submission: No temporary exposure keys to submit.", log: .api)
```

File: src/xcode/ENA/ENA/Source/Services/Exposure Submission/ExposureSubmissionService.swift [hunk 19]
```diff
@@ -150,7 +150,80 @@ class ENAExposureSubmissionService: ExposureSubmissionService {
...
+			completion(.preconditionError(.keysNotShared))
+			return
+		}
 
+		guard !keys.isEmpty || !checkins.isEmpty else {
+			Log.info("Cancelled submission: No temporary exposure keys or checkins to submit.", log: .api)
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/Services/Exposure Submission/ExposureSubmissionService.swift]:
```suggestion
			Log.info("Cancelled SRS exposure submission: No temporary exposure keys or checkins to submit.", log: .api)
```

File: src/xcode/ENA/ENA/Source/Services/Exposure Submission/ExposureSubmissionService.swift [hunk 20]
```diff
@@ -150,7 +150,80 @@ class ENAExposureSubmissionService: ExposureSubmissionService {
...
+			self.submitExposureCleanup(submissionTestType: .srs(submissionType))
+			return
+		}
+
+		// we need the app configuration first…
+		appConfigurationProvider
+			.appConfiguration()
+			.sink { [weak self] appConfig in
+				guard let self = self else {
+					Log.error("Failed to create string self")
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/Services/Exposure Submission/ExposureSubmissionService.swift]:
```suggestion
					Log.error("Failed to create strong self")
```

File: src/xcode/ENA/ENA/Source/Services/Exposure Submission/ExposureSubmissionService.swift [hunk 21]
```diff
@@ -312,7 +385,44 @@ class ENAExposureSubmissionService: ExposureSubmissionService {
...
+				Log.info("Successfully completed SRS exposure submission.", log: .api)
+				completion(nil)
+			case .failure(let error):
+				Log.error("Error while submitting SRS diagnosis keys: \(error.localizedDescription)", log: .api)
+				
+				completion(error)
+			}
+		}
+	}
+	/// Helper method that handles only the submission of the keys. Use this only if you really just want to do the
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/Services/Exposure Submission/ExposureSubmissionService.swift:426]: Can we have a space here? and a small addition
```suggestion
	
	/// Helper method that handles only the submission of the keys for regular submission. Use this only if you really just want to do the
```

File: src/xcode/ENA/ENA/Source/Services/Exposure Submission/ExposureSubmissionService.swift [hunk 22]
```diff
@@ -312,7 +385,44 @@ class ENAExposureSubmissionService: ExposureSubmissionService {
 	/// Helper method that handles only the submission of the keys. Use this only if you really just want to do the
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/Services/Exposure Submission/ExposureSubmissionService.swift]:
```suggestion
	/// Helper method that handles only the submission of the keys for SRS. Use this only if you really just want to do the
```

File: src/xcode/ENA/ENA/Source/Services/PPAccessControl/Model/SRSError.swift [hunk 23]
```diff
@@ -5,42 +5,50 @@
 import Foundation
 
 enum SRSError: Error {
+	case ppacError(PPACError)
+    case otpError(OTPError)
```
Comment (puneetmahali1) [inline src/xcode/ENA/ENA/Source/Services/PPAccessControl/Model/SRSError.swift:9]: It's not aligned!

File: src/xcode/ENA/ENA/Source/Services/Exposure Submission/ExposureSubmissionService.swift [hunk 24]
```diff
@@ -357,18 +467,28 @@ class ENAExposureSubmissionService: ExposureSubmissionService {
...
+				coronaTestService.antigenTest.value?.submissionTAN = nil
+			}
 
-		/// Deactivate deadman notification while submitted test is still present
-		deadmanNotificationManager.resetDeadmanNotification()
+			/// Deactivate deadman notification while submitted test is still present
+			deadmanNotificationManager.resetDeadmanNotification()
+
+		case .srs:
+			break
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/Services/Exposure Submission/ExposureSubmissionService.swift:495]: A comment here would be nice, as we don't store SRS, we don't need a cleanup something like this.

File: src/xcode/ENA/ENA/Resources/Localization/de.lproj/Localizable.strings (hunk 4 above)
Comment (flxschmidt) [reply src/xcode/ENA/ENA/Resources/Localization/de.lproj/Localizable.strings:502]: `ü`

File: src/xcode/ENA/ENA/Source/Services/OTP/OTPService.swift [hunk 25]
```diff
@@ -193,4 +217,37 @@ final class OTPService: OTPServiceProviding {
 			}
 		}
 	}
+	
+	private func authorizeSrs(_ otp: String, with ppacToken: PPACToken, completion: @escaping (Result<String, OTPError>) -> Void) {
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/Services/OTP/OTPService.swift]:
```suggestion
	private func authorizeSRS(_ otp: String, with ppacToken: PPACToken, completion: @escaping (Result<String, OTPError>) -> Void) {
```

File: src/xcode/ENA/ENA/Source/Services/SRSService/MockSrsService.swift [hunk 26]
```diff
@@ -0,0 +1,13 @@
...
+//
+
+import Foundation
+
+class MockSRSService: SRSServiceProviding {
+	func authenticate(completion: @escaping SRSAuthenticationResponse) {
+		completion(.success("TEST"))
+	}
+	
+
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/Services/SRSService/MockSrsService.swift]:
```suggestion
```

File: src/xcode/ENA/ENA/Source/Services/__tests__/Mocks/MockTestStore.swift [hunk 27]
```diff
@@ -83,12 +83,20 @@ final class MockTestStore: Store, PPAnalyticsData {
 	
 	var selectedLocalStatisticsRegions: [LocalStatisticsRegion] = []
 
+	// MARK: - SRS Providing
+
+	var ppacApiTokenSRS: TimestampedToken?
+	var previousPpacApiTokenSRS: TimestampedToken?
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/Services/__tests__/Mocks/MockTestStore.swift]:
```suggestion
	var previousPPACApiTokenSrs: TimestampedToken?
```

File: src/xcode/ENA/ENA/Source/Services/__tests__/Mocks/MockTestStore.swift [hunk 28]
```diff
@@ -83,12 +83,20 @@ final class MockTestStore: Store, PPAnalyticsData {
...
+	var otpTokenSRS: OTPToken?
+    var otpSrsAuthorizationDate: Date?
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/Services/__tests__/Mocks/MockTestStore.swift:91]:
```suggestion
    var otpSRSAuthorizationDate: Date?
```

File: src/xcode/ENA/ENA/Source/Services/SRSService/MockSrsService.swift (hunk 26 above)
Comment (puneetmahali1) [inline src/xcode/ENA/ENA/Source/Services/SRSService/MockSrsService.swift]: Can we remove one line space here?

File: src/xcode/ENA/ENA/Source/Services/__tests__/Mocks/MockTestStore.swift [hunk 29]
```diff
@@ -83,12 +83,20 @@ final class MockTestStore: Store, PPAnalyticsData {
 	
 	var selectedLocalStatisticsRegions: [LocalStatisticsRegion] = []
 
+	// MARK: - SRS Providing
+
+	var ppacApiTokenSRS: TimestampedToken?
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/Services/__tests__/Mocks/MockTestStore.swift]:
```suggestion
	var ppacApiTokenSrs: TimestampedToken?
```

File: src/xcode/ENA/ENA/Source/Services/__tests__/Mocks/MockTestStore.swift [hunk 30]
```diff
@@ -83,12 +83,20 @@ final class MockTestStore: Store, PPAnalyticsData {
...
+	var previousPpacApiTokenSRS: TimestampedToken?
+	var otpTokenSRS: OTPToken?
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/Services/__tests__/Mocks/MockTestStore.swift]:
```suggestion
	var otpTokenSrs: OTPToken?
```

File: src/xcode/ENA/ENA/Source/AppDelegate & Globals/AppDelegate.swift [hunk 31]
```diff
@@ -526,6 +526,16 @@ class AppDelegate: UIResponder, UIApplicationDelegate, CoronaWarnAppDelegate, Re
 		ppacService: ppacService,
 		otpService: otpService
 	)
+	
+	/// SRS Service singleton
```
Comment (flxschmidt) [inline src/xcode/ENA/ENA/Source/AppDelegate & Globals/AppDelegate.swift:530]: If you want to use a Singleton Pattern, you should better make the constructor private and use a static instance of the singleton inside of the Datatype. Example: 

```swift
class SRSService {
   static let shared: Self = .init()
   struct Config { // the dependencies you need in your singleton ... }

   class func setup(with config: Config) {
      Self.shared.config = config 
   }

   private init() {}
}
```

File: src/xcode/ENA/ENA/Source/HTTPClient/LocatorsNotYetUsed/KeySubmission.swift (hunk 6 above)
Comment (flxschmidt) [reply src/xcode/ENA/ENA/Source/HTTPClient/LocatorsNotYetUsed/KeySubmission.swift:24]: yeah great!

File: src/xcode/ENA/ENA/Source/Services/__tests__/Mocks/MockTestStore.swift (hunk 28 above)
Comment (puneetmahali1) [inline src/xcode/ENA/ENA/Source/Services/__tests__/Mocks/MockTestStore.swift:91]: should be aligned!

File: src/xcode/ENA/ENA/Source/Services/Exposure Submission/ExposureSubmissionService.swift [hunk 32]
```diff
@@ -312,7 +385,44 @@ class ENAExposureSubmissionService: ExposureSubmissionService {
...
+			checkins: checkins,
+			checkinProtectedReports: checkInProtectedReports,
+			tan: nil,
+			submissionType: submissionType.protobufType
+		)
+				
+		let resource = SRSKeySubmissionResource(payload: payload, srsOtp: srsOtp)
+		restServiceProvider.load(resource) { result in
+			
+			switch result {
```
Comment (flxschmidt) [inline src/xcode/ENA/ENA/Source/Services/Exposure Submission/ExposureSubmissionService.swift:412]: I love Result types!
Of course you could also have those completion parameter in 
`_submitSRS(completion(Result<Void, ServiceError<KeySubmissionResourceError>>))`

**Review by flxschmidt — APPROVED** [review]
Really love to read you code! <3

File: src/xcode/ENA.xcworkspace/xcshareddata/swiftpm/Package.resolved (hunk 1 above)
Comment (30mar) [reply src/xcode/ENA.xcworkspace/xcshareddata/swiftpm/Package.resolved]: this is already reverted

File: src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/ExposureSubmissionCoordinator.swift (hunk 14 above)
Comment (30mar) [reply src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/ExposureSubmissionCoordinator.swift:478]: indentation? yes it is correct in xcode

File: src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/ExposureSubmissionCoordinatorModel.swift (hunk 16 above)
Comment (30mar) [reply src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/ExposureSubmissionCoordinatorModel.swift]: no i fixed this

File: src/xcode/ENA/ENA/Source/Services/__tests__/Mocks/MockTestStore.swift (hunk 29 above)
Comment (30mar) [reply src/xcode/ENA/ENA/Source/Services/__tests__/Mocks/MockTestStore.swift]: Why should it be Srs not SRS, we are writing PPAC not Ppac for example

File: src/xcode/ENA/ENA/Source/Services/PPAccessControl/Model/SRSError.swift (hunk 23 above)
Comment (30mar) [reply src/xcode/ENA/ENA/Source/Services/PPAccessControl/Model/SRSError.swift:9]: it is in xcode, github issue

File: src/xcode/ENA/ENA/Source/AppDelegate & Globals/AppDelegate.swift (hunk 31 above)
Comment (30mar) [reply src/xcode/ENA/ENA/Source/AppDelegate & Globals/AppDelegate.swift:530]: it is not a singleton pattern it is just an initializer, like the proprieties above, i will remove the comment for confusion

File: src/xcode/ENA/ENA/Source/Services/Exposure Submission/ExposureSubmissionService.swift (hunk 32 above)
Comment (30mar) [reply src/xcode/ENA/ENA/Source/Services/Exposure Submission/ExposureSubmissionService.swift:412]: can you explain more, i didnt understand this

File: src/xcode/ENA/ENA/Source/Services/PPAccessControl/Model/SRSError.swift (hunk 23 above)
Comment (puneetmahali1) [reply src/xcode/ENA/ENA/Source/Services/PPAccessControl/Model/SRSError.swift:9]: alright!

File: src/xcode/ENA/ENA/Source/HTTPClient/Resources/KeySubmission/SRSKeySubmissionResource.swift (hunk 5 above)
Comment (puneetmahali1) [reply src/xcode/ENA/ENA/Source/HTTPClient/Resources/KeySubmission/SRSKeySubmissionResource.swift]: @30mar Did you remove the commented line?

**Review by puneetmahali1 — APPROVED** [review]

File: src/xcode/ENA/ENA/Source/Services/Exposure Submission/ExposureSubmissionService.swift (hunk 32 above)
Comment (flxschmidt) [reply src/xcode/ENA/ENA/Source/Services/Exposure Submission/ExposureSubmissionService.swift:412]: Your code is pretty fine. I just saw, that the completion handler expects an `Error?` type. That means `nil` for success context. Just my personal taste: 
`completion(.success(()))` looks more sexy than `completion(nil)`

File: src/xcode/ENA/ENA/Source/Workers/Store/Store.swift [hunk 33]
```diff
@@ -121,6 +125,17 @@ protocol LocalStatisticsCaching: AnyObject {
 	var selectedLocalStatisticsRegions: [LocalStatisticsRegion] { get set }
 }
 
+protocol SRSProviding: AnyObject {
+	/// PPAC token for Self-Report Submission (SRS)
+	var ppacApiTokenSRS: TimestampedToken? { get set }
+	/// Previous PPAC token for Self-Report Submission (SRS)
+	var previousPPACApiTokenSrs: TimestampedToken? { get set }
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/Workers/Store/Store.swift:132]: previousPpacApiTokenSrs

File: src/xcode/ENA/ENA/Source/Workers/Store/Store.swift [hunk 34]
```diff
@@ -121,6 +125,17 @@ protocol LocalStatisticsCaching: AnyObject {
 	var selectedLocalStatisticsRegions: [LocalStatisticsRegion] { get set }
 }
 
+protocol SRSProviding: AnyObject {
+	/// PPAC token for Self-Report Submission (SRS)
+	var ppacApiTokenSRS: TimestampedToken? { get set }
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/Workers/Store/Store.swift]:
```suggestion
	var ppacApiTokenSrs: TimestampedToken? { get set }
```

File: src/xcode/ENA/ENA/Source/Workers/Store/Store.swift [hunk 35]
```diff
@@ -121,6 +125,17 @@ protocol LocalStatisticsCaching: AnyObject {
...
+	/// Previous PPAC token for Self-Report Submission (SRS)
+	var previousPPACApiTokenSrs: TimestampedToken? { get set }
+	/// OTP for Self-Report Submission (SRS)
+	var otpTokenSRS: OTPToken? { get set }
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/Workers/Store/Store.swift]:
```suggestion
	var otpTokenSrs: OTPToken? { get set }
```

File: src/xcode/ENA/ENA/Source/Workers/Store/SecureStore.swift [hunk 36]
```diff
@@ -553,6 +573,29 @@ extension SecureStore: ErrorLogProviding {
 	#endif
 }
 
+extension SecureStore: SRSProviding {
+	
+	var ppacApiTokenSRS: TimestampedToken? {
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/Workers/Store/SecureStore.swift]:
```suggestion
	var ppacApiTokenSrs: TimestampedToken? {
```

**Review by naveeddotio — APPROVED** [review]

File: src/xcode/ENA/ENA/Resources/Localization/de.lproj/Localizable.strings [hunk 37]
```diff
@@ -745,6 +745,7 @@ Bei ausgeschalteter Hintergrundaktualisierung müssen Sie die App täglich aufru
 
 "SRS_ConsentScreen_Acknowledgement_2" = "Ihre Identität bleibt geheim. Andere Nutzer erfahren nicht, wer sein Testergebnis geteilt hat.";
 
+"SRS_ConsentScreen_instruction3" = "Sie können für einen Selbsttest kein offizielles COVID-Testzertifikat erhalten.";
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Resources/Localization/de.lproj/Localizable.strings]: @30mar please remove this, as this is extra
